src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2569,7 +2571,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           }
         }
         val statsTyped = typedStats(block.stats, context.owner, warnPure = false)
-        val expr1 = typed(block.expr, mode &~ (FUNmode | QUALmode), pt)
+        val expr1 = typed(block.expr, mode &~ (FUNmode | QUALmode | APPSELmode), pt)
```

Comment:
@retronym this change fixed the last example mentioned on the PR (`{ println("sara"); (new Foo).m}.m`)